### PR TITLE
Public ip metrics

### PR DESCRIPTION
**Pull request by iri989**

### CODE REVIEW

File: cosmic-metrics-collector/src/main/java/com/github/missioncriticalcloud/cosmic/metricscollector/repositories/PublicIPAddressMetricsRepository.java [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package com.github.missioncriticalcloud.cosmic.metricscollector.repositories;
+
+/**
```
Comment (bheuvel) [inline cosmic-metrics-collector/src/main/java/com/github/missioncriticalcloud/cosmic/metricscollector/repositories/PublicIPAddressMetricsRepository.java]: Suggest to remove these 'created by' comments? (Other files as well)

File: cosmic-metrics-collector/src/main/resources/queries.yml [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    SELECT
+          d.uuid domainUuid, a.uuid accountUuid, ip.uuid resourceUuid,
+          ip.state state
+        FROM user_ip_address ip
+        JOIN domain d
+          ON d.id = ip.domain_id
+        JOIN account a
+          ON a.id = ip.account_id
+        WHERE
+          ip.removed IS NULL
```
Comment (bheuvel) [inline cosmic-metrics-collector/src/main/resources/queries.yml:31]: Newline missing

File: cosmic-metrics-collector/src/test/resources/test-schema.sql [hunk 3]
```diff
@@ -130,3 +130,52 @@ CREATE TABLE IF NOT EXISTS `service_offering` (
...
+  KEY `i_user_ip_address__source_nat` (`source_nat`)
+--   CONSTRAINT `fk_user_ip_address__account_id` FOREIGN KEY (`account_id`) REFERENCES `account` (`id`),
+--   CONSTRAINT `fk_user_ip_address__data_center_id` FOREIGN KEY (`data_center_id`) REFERENCES `data_center` (`id`) ON DELETE CASCADE,
+--   CONSTRAINT `fk_user_ip_address__network_id` FOREIGN KEY (`network_id`) REFERENCES `networks` (`id`),
+--   CONSTRAINT `fk_user_ip_address__physical_network_id` FOREIGN KEY (`physical_network_id`) REFERENCES `physical_network` (`id`) ON DELETE CASCADE,
+--   CONSTRAINT `fk_user_ip_address__source_network_id` FOREIGN KEY (`source_network_id`) REFERENCES `networks` (`id`),
+--   CONSTRAINT `fk_user_ip_address__vlan_db_id` FOREIGN KEY (`vlan_db_id`) REFERENCES `vlan` (`id`) ON DELETE CASCADE,
+--   CONSTRAINT `fk_user_ip_address__vm_id` FOREIGN KEY (`vm_id`) REFERENCES `vm_instance` (`id`),
+--   CONSTRAINT `fk_user_ip_address__vpc_id` FOREIGN KEY (`vpc_id`) REFERENCES `vpc` (`id`) ON DELETE CASCADE
+) ENGINE=InnoDB AUTO_INCREMENT=50 DEFAULT CHARSET=utf8;
```
Comment (eliasgomes) [inline cosmic-metrics-collector/src/test/resources/test-schema.sql:181]: New line missing.

File: cosmic-metrics-collector/src/test/resources/test-data.sql [hunk 4]
```diff
@@ -17,3 +17,8 @@ INSERT INTO `service_offering`
   (id, cpu, ram_size)
 VALUES
   (1, 2, 4096);
+
```
Comment (eliasgomes) [inline cosmic-metrics-collector/src/test/resources/test-data.sql]: I'd suggest moving this new test data to a new file. It keeps things easier to maintain as the test cases grow.

File: cosmic-metrics-collector/src/main/java/com/github/missioncriticalcloud/cosmic/metricscollector/repositories/jdbc/PublicIPAddressMetricsJdbcRepository.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+import org.springframework.jdbc.core.namedparam.NamedParameterJdbcTemplate;
+import org.springframework.stereotype.Repository;
+
+import java.util.List;
+import java.util.Properties;
+
+@Repository
+public class PublicIPAddressMetricsJdbcRepository implements PublicIPAddressMetricsRepository {
```
Comment (eliasgomes) [inline cosmic-metrics-collector/src/main/java/com/github/missioncriticalcloud/cosmic/metricscollector/repositories/jdbc/PublicIPAddressMetricsJdbcRepository.java]: I suggest to use the _spring-like_ camel-case for names wherever it applies. For example: `PublicIpAddressMetricsJdbcRepository`.

File: cosmic-metrics-collector/src/main/java/com/github/missioncriticalcloud/cosmic/metricscollector/tasks/PublicIPAddressMetricCollector.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.stereotype.Component;
+import org.springframework.util.StopWatch;
+
+import java.util.List;
+import java.util.logging.Logger;
+
+@Component
+public class PublicIPAddressMetricCollector implements MetricCollector{
```
Comment (eliasgomes) [inline cosmic-metrics-collector/src/main/java/com/github/missioncriticalcloud/cosmic/metricscollector/tasks/PublicIPAddressMetricCollector.java]: Please add a space between `MetricCollector` and `{` for better reading of the code.

**Review by eliasgomes — CHANGES_REQUESTED** [review]
I added some remarks to this pull request. Please also properly adjust the commits and commit messages so we can easily revert and identify them, if necessary.

File: cosmic-metrics-collector/src/main/resources/queries.yml [hunk 7]
```diff
@@ -16,3 +16,16 @@ virtual-machine-metrics-repository:
       vm.state IN ('Running', 'Stopped')
     AND
       vm.removed IS NULL
+
+public-ip-address-metrics-repository:
+  get-metrics: |
+    SELECT
```
Comment (eliasgomes) [inline cosmic-metrics-collector/src/main/resources/queries.yml:22]: We should filter by `state` as well. If there's a problem while the public IP address is being allocated (with `Allocating` state), we shouldn't bill that IP. So, please add to the `WHERE` clause: `ip.state = 'Allocated'`

**Review by eliasgomes — CHANGES_REQUESTED** [review]

**Review by eliasgomes — APPROVED** [review]

**Pull request merged**